sw/device/sca/aes_serial/aes_serial.c

```diff
@@ -0,0 +1,293 @@
+ *
+ * @param from input value in binary format.
+ * @param to   output value in hex format.
+ *
+ * @return SIMPLE_SERIAL_OK on success, SIMPLE_SERIAL_ERROR otherwise.
+ */
+static uint8_t hex_value(char from, char *to) {
+  if (from >= '0' && from <= '9') {
+    *to = from - '0';
+  } else if (from >= 'A' && from <= 'F') {
+    *to = from - 'A' + 10;
+  } else if (from >= 'a' && from <= 'f') {
+    *to = from - 'a' + 10;
+  } else {
+    return SIMPLE_SERIAL_ERROR;
+  }
+  return SIMPLE_SERIAL_OK;
+}
+
+/**
+ * Convert `num` bytes `from` hex `to` binary format.
+ *
+ * `from` size is expected to by twice as big as `to`.
+ *
+ * @param from input buffer for hex formatted characters.
+ * @param to   output binary buffer.
+ * @param num  number of characters in output buffer.
+ *
+ * @return SIMPLE_SERIAL_OK on success, SIMPLE_SERIAL_ERROR otherwise.
+ */
+static uint8_t a2b_hex(const char *from, char *to, size_t num) {
+  for (int i = 0; i < num; ++i) {
+    char hi, lo;
+    if (hex_value(from[i * 2], &hi) || hex_value(from[i * 2 + 1], &lo)) {
+      return SIMPLE_SERIAL_ERROR;
+    }
+    to[i] = ((hi & 0xFF) << 4) | (lo & 0xFF);
+  }
+  return SIMPLE_SERIAL_OK;
+}
+
+/**
+ * Send `data` to UART in hex byte format.
+ */
+static void print_hex_buffer(const char *data, size_t data_len) {
+  static const char b2a_hex_values[16] = {'0', '1', '2', '3', '4', '5',
+                                          '6', '7', '8', '9', 'A', 'B',
+                                          'C', 'D', 'E', 'F'};
+  for (int i = 0; i < data_len; ++i) {
+    uart_send_char(b2a_hex_values[data[i] >> 4]);
+    uart_send_char(b2a_hex_values[data[i] & 0xF]);
+  }
+}
+
+/**
+ * Send simple serial command response via UART.
+ *
+ * @param cmd_tag  Simple serial command tag.
+ * @param data     Response data. Converted to hex format by this function.
+ * @param data_len data length.
+ */
+static void print_cmd_response(const char cmd_tag, const char *data,
+                               size_t data_len) {
+  uart_send_char(cmd_tag);
+  print_hex_buffer(data, data_len);
+  uart_send_char('\n');
+}
+
+/**
+ * Return number of AES blocks from `plain_text_len`.
+ */
+static size_t get_num_blocks(size_t plain_text_len) {
+  size_t num_blocks = plain_text_len / AES_BLOCK_SIZE;
+  if ((plain_text_len - (num_blocks * AES_BLOCK_SIZE)) > 0) {
+    ++num_blocks;
+  }
+  return num_blocks;
+}
+
+/**
+ * Simple serial set AES key command.
+ *
+ * @param aes_cfg AES configuration object.
+ * @param key     AES key.
+ * @param key_len AES key length.
+ *
+ * @return SIMPLE_SERIAL_OK on success, SIMPLE_SERIAL_ERROR otherwise.
+ */
+static uint8_t simple_serial_set_key(const aes_cfg_t *aes_cfg, const char *key,
+                                     size_t key_len) {
+  if (key_len != AES_KEY_SIZE) {
+    return SIMPLE_SERIAL_ERROR;
+  }
+  aes_clear();
+  while (!aes_idle()) {
+    ;
+  }
+  aes_key_put((const void *)key, aes_cfg->key_len);
+  aes_init(*aes_cfg);
```

Comment:
Please note that this sequence should be switched when rebasing on current master.

It would probably still work as we operate in manual mode here. But it would be more clean to first write the config and then the key. I had to change to design to require the initialization always before writing the key and IV in automatic/normal mode.